api/handle_ingestion.go

```diff
@@ -33,9 +34,60 @@ func handleIngestion(uc usecases.Usecases) func(c *gin.Context) {
 
 		usecase := usecasesWithCreds(ctx, uc).NewIngestionUseCase()
 		nb, err := usecase.IngestObject(ctx, organizationId, objectType, objectBody)
+		var validationError models.IngestionValidationErrors
+		if errors.As(err, &validationError) {
+			_, objectErr := validationError.GetSomeItem()
+			c.JSON(http.StatusBadRequest, dto.APIErrorResponse{
+				Message:   "Input validation error",
+				Details:   objectErr,
+				ErrorCode: dto.SchemaMismatchError,
+			})
+			return
+		} else if presentError(ctx, c, err) {
+			return
+		}
+		if nb == 0 {
+			c.Status(http.StatusOK)
+			return
+		}
+		c.Status(http.StatusCreated)
+	}
+}
+
+func presentIngestionValidationError(c *gin.Context, err error) bool {
+	var validationError models.IngestionValidationErrors
+	if errors.As(err, &validationError) {
+		_, objectErr := validationError.GetSomeItem()
+		c.JSON(http.StatusBadRequest, dto.APIErrorResponse{
+			Message:   "Input validation error",
+			Details:   objectErr,
+			ErrorCode: dto.SchemaMismatchError,
+		})
+		return true
+	}
+	return false
+}
+
+func handleIngestionPartialUpsert(uc usecases.Usecases) func(c *gin.Context) {
+	return func(c *gin.Context) {
+		ctx := c.Request.Context()
+		organizationId, err := utils.OrganizationIdFromRequest(c.Request)
 		if presentError(ctx, c, err) {
 			return
 		}
+
+		objectType := c.Param("object_type")
+		objectBody, err := io.ReadAll(c.Request.Body)
```

Comment:
Ideally, we would pass the `io.Reader` around instead of converting it to a byte slice here, but I noticed we use `gjson` later on, which does not support parsing from readers...

I would personally either unmarshal it to a `json.RawMessage` here, in order for JSON errors to fail from within the handler. Whether doing this or not, I think it would be wise to use `ReadFull` into a preallocated buffer of `c.Request.ContentLength` size to avoid continuously reallocating the buffer (especially in the "ingest all" endpoint, whose size could become quite large).

That discussion could be had offline, but we might want to discuss how we ingest efficiently several objects when the size is unknown, because the strategy of loading everything into memory might come back to haunt us at scale.